Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp

```diff
@@ -404,8 +404,16 @@ void WebCookieJar::setCookieAsync(WebCore::Document& document, const URL& url, c
 }
 
 #if HAVE(COOKIE_CHANGE_LISTENER_API)
-void WebCookieJar::addChangeListener(const String& host, const WebCore::CookieChangeListener& listener)
+void WebCookieJar::addChangeListener(const WebCore::Document& document, const String& host, const WebCore::CookieChangeListener& listener)
 {
+    RefPtr webFrame = document.frame() ? WebFrame::fromCoreFrame(*document.protectedFrame()) : nullptr;
+    if (!webFrame || !webFrame->page())
+        return;
+
+    ApplyTrackingPrevention applyTrackingPreventionInNetworkProcess = ApplyTrackingPrevention::No;
+    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), document.cookieURL(), applyTrackingPreventionInNetworkProcess))
```

Comment:
@szewai probably remembers the details but I think we need a similar check on the network process side.

See for example this code running in the network process when `document.cookie` is called:
```
RetainPtr<NSArray> NetworkStorageSession::cookiesForURL(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) const
{
    if (applyTrackingPrevention == ApplyTrackingPrevention::Yes && shouldBlockCookies(firstParty, url, frameID, pageID, shouldRelaxThirdPartyCookieBlocking))
        return nil;
    return httpCookiesForURL(cookieStorage().get(), firstParty, sameSiteInfo, url);
}
```

It checks shouldBlockCookies() too.

My bet is that we need a similar check on network process side when registering a cookie listener.

Reply:
Since a compromised WebContent process can skip all these checks, a check in the Network process is essential.

Probably the check belongs inside the handler for SubscribeToCookieChangeNotifications.

Reply:
To be clear, I'm not encouraging to move the check to the network process but rather add an additional one in the NetworkProcess, just like we do for the `document.cookie` code path. I think we should be consistent.

Reply:
@szewai I've added the Network Process checks to the `addChangeListener` code path. Do we also want them in the `removeChangeListener` path? I thought about doing this, but it didn't seem feasible in some places. For example, where would `WebCookieCache::clear()` or `NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess()` get the needed document?

Also, if the WebProcess check passes, but the Network Process check fails, we would have already added the listener to `WebCookieaJar::m_changeListeners`. Is this ok or does the Network Process need to IPC back to let WebProcess know if it's ok to add the listener to the map. The same issue exists with the two checks in the Network Process. If the check in NetworkConnectionToWebProcess passes but the check the NetworkStorageSession fails, we've already added the listener to `NetworkConnectionToWebProcess ::m_hostsWithCookieListeners`. Is this ok?

Reply:
> Do we also want them in the removeChangeListener path?

We don't need to perform check for `removeChangeListener` since it does not give web process more access it does not have. 

> Also, if the WebProcess check passes, but the Network Process check fails, we would have already added the listener to WebCookieaJar::m_changeListeners. Is this ok or does the Network Process need to IPC back to let WebProcess know if it's ok to add the listener to the map?

With existing implementation, technically web process needs to know whether the subscription succeeds or not, because it sends only one message for one host.

> The same issue exists with the two checks in the Network Process. If the check in NetworkConnectionToWebProcess passes but the check the NetworkStorageSession fails, we've already added the listener to NetworkConnectionToWebProcess ::m_hostsWithCookieListeners. Is this ok?

`NetworkConnectionToWebProcess::m_hostsWithCookieListeners` looks unnecessary to me, since `NetworkStorageSession` already tracks the information. Maybe we can just drop it.

Reply:
> We don't need to perform check for removeChangeListener since it does not give web process more access it does not already have.
Having equivalent checks for `removeChangeListener` would prevent a web process from removing cookie listeners it does not own. No idea why that’s something a compromised web process would want to do, so maybe it's not important.

Reply:
Cookie change observer in network process is stored by connection, so a WP cannot change the subscription state of another WP.

Reply:
I've added the functionality where the Network Process returns a boolean telling the Web Process if the checks passed and listener was indeed added or not. 

It seems that `NetworkConnectionToWebProcess::m_hostsWithCookieListeners` is passed to `NetworkStorageSession::stopListeningForCookieChangeNotifications` to stop the listening. This seems important,  so I didn't remove that map yet. I've updated the code to only add the host to the map if `shouldBlockCookies` in `NetworkStorageSession` is not true. What do we think of that?